modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java

```diff
@@ -0,0 +1,588 @@
+        assertArrayEquals(copyOfRange(randomBytes, 2 * 1024, 3 * 1024), toByteArray(buffer));
+        assertEquals(3 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.readFully(buffer.rewind()));
+        assertArrayEquals(copyOfRange(randomBytes, 3 * 1024, 4 * 1024), toByteArray(buffer));
+        assertEquals(4 * 1024, fileIo.position());
+
+        assertEquals(-1, fileIo.readFully(buffer.rewind()));
+        assertEquals(4 * 1024, fileIo.position());
+    }
+
+    @Test
+    void testReadByPositionFully() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        writeBytes(testFilePath, randomBytes);
+
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.readFully(buffer.rewind(), 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 1024, 2 * 1024), toByteArray(buffer));
+        assertEquals(0, fileIo.position());
+
+        assertEquals(1024, fileIo.readFully(buffer.rewind(), 2 * 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 2 * 1024, 3 * 1024), toByteArray(buffer));
+        assertEquals(0, fileIo.position());
+
+        assertEquals(-1, fileIo.readFully(buffer.rewind(), 4 * 1024));
+        assertEquals(0, fileIo.position());
+    }
+
+    @Test
+    void testReadToByteArrayFully() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        writeBytes(testFilePath, randomBytes);
+
+        byte[] buffer = new byte[1024];
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.readFully(buffer, 0, 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 0, 1024), buffer);
+        assertEquals(1024, fileIo.position());
+
+        assertEquals(1024, fileIo.readFully(buffer, 0, 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 1024, 2 * 1024), buffer);
+        assertEquals(2 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.readFully(buffer, 0, 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 2 * 1024, 3 * 1024), buffer);
+        assertEquals(3 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.readFully(buffer, 0, 1024));
+        assertArrayEquals(copyOfRange(randomBytes, 3 * 1024, 4 * 1024), buffer);
+        assertEquals(4 * 1024, fileIo.position());
+
+        assertEquals(-1, fileIo.readFully(buffer, 0, 1024));
+        assertEquals(4 * 1024, fileIo.position());
+    }
+
+    @Test
+    void testWrite() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.write(ByteBuffer.wrap(copyOfRange(randomBytes, 0, 1024))));
```

Comment:
Can we also test slices? Also, this piece of code can be extracted in to a method, like "ByteBuffer rangeBuffer(arr, from, to)"

Reply:
Modify it